feat: Support function call requests/responses

caffeinism opened the pull request:
I have implemented the issue (https://github.com/samuelint/langchain-openai-api-bridge/issues/50)

I also changed the OpenAI related messages internally from BaseModel to the actual internal data format of openai python sdk. This shouldn't be an issue in the data format actually used by users, since we only added tools and tool_choice to CreateAgentDto.

In the chat completion unit test, message.role = "ai" was incorrect for the openai message format, so i changed it to "assistant".

---

### Example:

Server
```python
class AgentFactory(BaseAgentFactory):
    async def create_agent(self, dto: CreateAgentDto):
        llm = self.create_llm(
            ...
        )
        return llm.bind_tools(dto.tools)
```

Client
```python
tools = [{
    "type": "function",
    "function": {
        "name": "get_weather",
        "description": "Get current temperature for a given location.",
        "parameters": {
            "type": "object",
            "properties": {
                "location": {
                    "type": "string",
                    "description": "City and country e.g. Bogotá, Colombia"
                }
            },
            "required": [
                "location"
            ],
            "additionalProperties": False
        },
        "strict": True
    }
}]

chat_completion = await openai_client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {
            "role": "user",
            "content": 'What is the weather like in London today?',
        }
    ],
    tools=tools,
)

# ChatCompletion(id='...', choices=[Choice(finish_reason='tool_calls', index=0, logprobs=None, message=ChatCompletionMessage(content=None, refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=[ChatCompletionMessageToolCall(id='...', function=Function(arguments='{"location": "London, UK"}', name='get_weather'), type='function')]))], created=1751731311, model='gpt-4o-mini', object='chat.completion', service_tier=None, system_fingerprint='', usage=CompletionUsage(completion_tokens=-1, prompt_tokens=-1, total_tokens=-1, completion_tokens_details=None, prompt_tokens_details=None))
```

caffeinism committed:
feat: Support function call requests/responses

caffeinism committed:
fix: Allow finish_reason=tool_call in streaming mode

samuelint commented on langchain_openai_api_bridge/chat_completion/chat_completion_chunk_choice_adapter.py:
```diff
@@ -4,33 +4,37 @@
+    else:
+        function_call = None
+
+    return ChoiceDelta(
```
The related tests appear to have only been updated for existing code, not for the newly introduced logic.
Tests for the new logic should also be added in `tests/test_unit/chat_completion/test_chat_completion_chunk_choice_adapter.py`.

samuelint commented on langchain_openai_api_bridge/chat_completion/langchain_stream_adapter.py:
```diff
@@ -23,22 +21,26 @@ async def ato_chat_completion_chunk_stream(
                     system_fingerprint=self.system_fingerprint,
                 )
+                yield chat_completion_chunk
+                is_function_call = is_function_call or any(choice.delta.function_call for choice in chat_completion_chunk.choices)
```
The related tests appear to have only been updated for existing code, not for the newly introduced logic.
Tests for the new logic should also be added.

samuelint commented on langchain_openai_api_bridge/chat_completion/langchain_invoke_adapter.py:
```diff
@@ -1,56 +1,43 @@
-    def to_chat_completion_object(self, invoke_result) -> OpenAIChatCompletionObject:
-        message = self.__create_openai_chat_message(invoke_result)
-        id = self.__get_id(invoke_result)
+    def to_chat_completion_object(self, invoke_result: Output) -> ChatCompletion:
```
It would be great to include unit tests for this new logic.
The original code appears to lack test coverage, which was an oversight. At a minimum, the new logic should be covered by tests.

samuelint reviewed (changes requested):
A functional test is missing to verify that the new feature works as intended.
Adding one would help ensure the feature behaves correctly and prevent regressions in future changes.

samuelint commented:
For tests to pass in the pipeline, please have secrets set in your GitHub pipeline.

`ANTHROPIC_API_KEY `
`GROQ_API_KEY `
`OPENAI_API_KEY `

<img width="851" alt="image" src="https://github.com/user-attachments/assets/837ba80e-e29d-4d35-a904-b6be32b03678" />

caffeinism commented:
> For tests to pass in the pipeline, please have secrets set in your GitHub pipeline.
> 
> 
> 
> `ANTHROPIC_API_KEY `
> 
> `GROQ_API_KEY `
> 
> `OPENAI_API_KEY `
> 
> 
> 
> <img width="851" alt="image" src="https://github.com/user-attachments/assets/837ba80e-e29d-4d35-a904-b6be32b03678" />
> 
> 

Oh, so that should be added to my repository?

caffeinism committed:
test: Add chat completion chunk tests

samuelint commented:
> > For tests to pass in the pipeline, please have secrets set in your GitHub pipeline.
> > `ANTHROPIC_API_KEY `
> > `GROQ_API_KEY `
> > `OPENAI_API_KEY `
> > <img alt="image" width="851" src="https://private-user-images.githubusercontent.com/5473142/462927267-837ba80e-e29d-4d35-a904-b6be32b03678.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE4MTQ1NTIsIm5iZiI6MTc1MTgxNDI1MiwicGF0aCI6Ii81NDczMTQyLzQ2MjkyNzI2Ny04MzdiYTgwZS1lMjlkLTRkMzUtYTkwNC1iNmJlMzJiMDM2NzgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MDZUMTUwNDEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWM3Njk3MDBjOWJiMGFhNjA0YTE0ZGYxYWVjMmJmYjI5NzM4ZTQ1NjhiOTA2OTg0YzNlNDczMTQyZjI3Y2I5YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.E6pXMyYyBgBDqIQJt-LF2Pbw2EtgY3xfQXXUFMzpNbI">
> 
> Oh, so that should be added to my repository?

Yes, since the project is forked, it should be set in your repository. This is a known limitation on GitHub, where my secrets cannot be used in forked pull-requests.

caffeinism commented:
> > > For tests to pass in the pipeline, please have secrets set in your GitHub pipeline.
> > > `ANTHROPIC_API_KEY `
> > > `GROQ_API_KEY `
> > > `OPENAI_API_KEY `
> > > <img alt="image" width="851" src="https://private-user-images.githubusercontent.com/5473142/462927267-837ba80e-e29d-4d35-a904-b6be32b03678.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE4MTQ1NTIsIm5iZiI6MTc1MTgxNDI1MiwicGF0aCI6Ii81NDczMTQyLzQ2MjkyNzI2Ny04MzdiYTgwZS1lMjlkLTRkMzUtYTkwNC1iNmJlMzJiMDM2NzgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MDZUMTUwNDEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWM3Njk3MDBjOWJiMGFhNjA0YTE0ZGYxYWVjMmJmYjI5NzM4ZTQ1NjhiOTA2OTg0YzNlNDczMTQyZjI3Y2I5YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.E6pXMyYyBgBDqIQJt-LF2Pbw2EtgY3xfQXXUFMzpNbI">
> > 
> > 
> > Oh, so that should be added to my repository?
> 
> Yes, since the project is forked, it should be set in your repository. This is a known limitation on GitHub, where my secrets cannot be used in forked pull-requests.

For tests, what format should I add tests in? I'm not very familiar with something like this. In functional tests, I'm thinking of doing simple tool calls like addition to verify behavior, but for unit tests, I'm not sure how I should write them down. Should I just add the various tests in the format they're currently added in (I've added one commit as an example)?

For repository secrets, I've added them to my own repository, but it's a bit confusing to me that secrets from my own repository are used for actions that run on PRs here.

caffeinism committed:
fix: Change to only include role=assistant in the first chunk, the same as OpenAI's server.

caffeinism committed:
test: Add chat completion function call functional test

caffeinism commented:
I added a simple unittest and functional test for the function call functionality. I also found where the behavior of the server differs from the actual openai server, so I changed it (only the first chunk should contain role)

samuelint commented on tests/test_unit/chat_completion/test_chat_completion_chunk_choice_adapter.py:
```diff
@@ -93,18 +94,18 @@ def test_delta_message_have_args_role(self):
+        assert result.delta.role == "assistant"
 
-    def test_message_have_assistant_role_by_default(self):
+    def test_delta_message_have_none_role_by_default(self):
```
This test was there for a reason. It might break something, I think the behaviour should remains as before except if you have a good reason for it.

samuelint commented on tests/test_unit/chat_completion/test_chat_completion_chunk_choice_adapter.py:
```diff
@@ -195,15 +196,30 @@ def test_delta_message_have_args_role(self):
+        assert result.choices[0].delta.role == "assistant"
 
-    def test_message_have_assistant_role_by_default(self):
+    def test_delta_message_have_none_role_by_default(self):
```
This test was there for a reason. It might break something, I think the behaviour should remains as before except if you have a good reason for it.

caffeinism replied:
If so, is it okay to leave “assistant” as the default, but in the first actual behavior, only send “assistant” for the first chunk and None for the rest?

caffeinism replied:
This change is due to:
> I also found where the behavior of the server differs from the actual openai server, so I changed it (only the first chunk should contain role)

When returning a chunk, the openai server returns None by default, and only the very first chunk returns role.

samuelint replied:
It's ok if this is the default openai api behaviour.

Please add a unit test in the  `langchain_openai_api_bridge/chat_completion/langchain_stream_adapter.py` unit test.

Have a test function name that explain that behaviour, "Only the first chunk has role assistant, afterward role is None". That will document the use case.

samuelint commented on tests/test_unit/chat_completion/test_chat_completion_chunk_choice_adapter.py:
```diff
@@ -195,15 +196,30 @@ def test_delta_message_have_args_role(self):
+        result = to_openai_chat_completion_chunk_object(event)
+
+        assert result.choices[0].delta.function_call is not None
+        assert result.choices[0].delta.function_call.name == "my_func"
```
The test does not validate the args property. I think it should have a test for the name and another for the args

samuelint commented on tests/test_functional/fastapi_chat_completion_openai_function_call/test_server_openai_function_call.py:
```diff
@@ -0,0 +1,147 @@
+
+
+def test_chat_completion_function_call_weather_stream(openai_client: OpenAI):
+    tools = [{
```
Since tools section is the same in each tests, it could be re-used instead of copy/paste

samuelint reviewed (changes requested):
Great job, one last minor comment and it will be ready to be merged.

samuelint commented:
> > > > For tests to pass in the pipeline, please have secrets set in your GitHub pipeline.
> > > > `ANTHROPIC_API_KEY `
> > > > `GROQ_API_KEY `
> > > > `OPENAI_API_KEY `
> > > > <img alt="image" width="851" src="https://private-user-images.githubusercontent.com/5473142/462927267-837ba80e-e29d-4d35-a904-b6be32b03678.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE4MTQ1NTIsIm5iZiI6MTc1MTgxNDI1MiwicGF0aCI6Ii81NDczMTQyLzQ2MjkyNzI2Ny04MzdiYTgwZS1lMjlkLTRkMzUtYTkwNC1iNmJlMzJiMDM2NzgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MDZUMTUwNDEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWM3Njk3MDBjOWJiMGFhNjA0YTE0ZGYxYWVjMmJmYjI5NzM4ZTQ1NjhiOTA2OTg0YzNlNDczMTQyZjI3Y2I5YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.E6pXMyYyBgBDqIQJt-LF2Pbw2EtgY3xfQXXUFMzpNbI">
> > > 
> > > 
> > > Oh, so that should be added to my repository?
> > 
> > 
> > Yes, since the project is forked, it should be set in your repository. This is a known limitation on GitHub, where my secrets cannot be used in forked pull-requests.
> 
> For tests, what format should I add tests in? I'm not very familiar with something like this. In functional tests, I'm thinking of doing simple tool calls like addition to verify behavior, but for unit tests, I'm not sure how I should write them down. Should I just add the various tests in the format they're currently added in (I've added one commit as an example)?
> 
> For repository secrets, I've added them to my own repository, but it's a bit confusing to me that secrets from my own repository are used for actions that run on PRs here.

What you did is fine, you've respected how other tests are made. The goal of functional tests is to validate the feature work from the API standpoint.

caffeinism committed:
fix: Change to respond only to the finish_reason of the last stream

caffeinism committed:
test: Remove duplicate code

caffeinism committed:
test: Add arguments validation

caffeinism committed:
fix: Check attribute for Any type chunk

caffeinism committed:
test: Add stream validation

caffeinism commented:
I've add it
